src/main/java/com/chargerrobotics/utils/ArduinoSerial.java

```diff
@@ -85,20 +121,27 @@ private int receiveData() {
 			cs.reset();
 			int expectedChecksum = 0;
 			int expectedLength = 0;
+			byte[] receivedHeader = new byte[2];
+			ArduinoListener listener = null;
 			while (true) {
```

Comment:
What stops this method from blocking forever if one of the ports sends bad data or is disconnected?

Reply:
Unless I missed somewhere, it has a 20ms timeout if the packet is invalid. If the port is disconnected, it catches the SerialPortIOException and closes the port, then periodically tries to reopen it.